packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+#### UI Blocks
+
+Higher-level components, which usually offer fewer props compared to UI Primitives and contain translated text. These components use other WordPress technologies and are ready to be used in Gutenberg's UI.
+
+#### Utilities
+
+Components and other utilities function that don't necessarily render a piece of UI to screen, but are instead useful when implementing a given piece of functionality.
+
+### Components composition
+
+[To be expanded] E.g.:
+
+- Polymorphic Components
+- Using `children` vs custom render props vs arbitrary "data" props
+- Controlled and semi-controlled components
+- Composition patterns
+- Sub-components naming conventions
+- Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own)
+- ...
+
+### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`), `has*` (e.g. `hasValue`) or `enable*` (e.g. `enableScroll`)
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- ...
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group);
+
+#### TypeScript
+
+All new components should be written in TypeScript and should be typed using the `PolymorphicComponent` type (more details about polymorphism can be found above in the "Components composition" section).
+
+#### Emotion
+
+All new component should be styled using [Emotion](https://emotion.sh/docs/introduction).
+
+Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
+
+#### Context system
+
+[To be expanded]
+
+#### Hooks vs Components
+
+[To be expanded]
+
+#### Unit tests
+
+All new components should include unit tests using [testing library](https://testing-library.com/). Ideally, the component's main piece of functionality would be tested explicitely (rather than using snapshots).
+
+#### Storybook
+
+All new components should include Storybook examples.
+
+#### Documentation
+
+All components, in addition to being typed, should be using JSDoc when necessary.
+
+Each component that is exported from the `@wordpress/components` package should include a `README.md` file, explaining how to use the component, showing examples, and documenting all the props.
+
+#### Folder structure
+
+Following all previous guidelines, all new components should follow this folder structure:
```

Comment:
Does this structure apply to UI-ess utilities as well? As @gziolo mentioned above, it doesn't seem like these utilities need to be exposed as they are reusable implementation details rather than components to be consumed externally. If so, I'd lean towards separating utilities in the folder structure.

Reply:
> it doesn't seem like these utilities need to be exposed as they are reusable implementation details rather than components to be consumed externally

That's a good point! I guess it greatly depends on what we define as a "utility" — see for example this [draft list](https://github.com/WordPress/gutenberg/issues/33111#issuecomment-871569331) that @sarayourfriend and I worked on some time ago.

If we were to go with this list, I would argue that some of these "utilities" could be useful if exported.

> I'd lean towards separating utilities in the folder structure.

If you refer to a component's folder structure, then I agree that we could change the wording so that the suggested folder structure _does not_ apply to internal utility components/function.

With regards to the overall package folder structure, I left a [note](https://github.com/WordPress/gutenberg/pull/33960/files#r687024506) in the `CONTRIBUTING.md` file in this PR about it — would love to hear your opinion about it too!